File: updated_code/code/python/1704-determine-if-string-halves-are-alike.py
import string
import set

class HalvesAreAlike:
    """
    Determine if the halves of a given string are alike.

    A string is "alike" if it has the same number of vowels.
    """

    def __init__(self):
        """
        Initialize the solution.
        """

    def halves_are_alike(self, s: str) -> bool:
        """
        Determine if the halves of a given string are alike.

        Args:
        s (str): The input string.

        Returns:
        bool: Whether the halves of the string are alike.
        """

        vowels = set(string.ascii_lowercase + string.ascii_uppercase)
        
        half = len(s) // 2
        
        a = sum(1 for char in set(s[:half]).intersection(vowels))
        b = sum(1 for char in set(s[half:]).intersection(vowels))
        
        return a == b


# Test cases
solution = HalvesAreAlike()
print(solution.halves_are_alike("book"))  # True
print(solution.halves_are_alike("textbook"))  # False

File: updated_code/code/java/1704-determine-if-string-halves-are-alike.java
package com.example;

import java.util.Arrays;

/**
 * This class is used to determine if two halves of a string are alike.
 */
public class Solution {
    public boolean halvesAreAlike(String s) {
        // Check for empty string
        if (s.isEmpty()) {
            return true;
        }

        // Calculate the mid index
        int mid = s.length() / 2;

        // Split the string into two halves
        String half1 = s.substring(0, mid);
        String half2 = s.substring(mid);

        // Convert the strings to character arrays
        char[] chars1 = half1.toCharArray();
        char[] chars2 = half2.toCharArray();

        // Initialize the count of vowels in each half
        int count1 = 0;
        int count2 = 0;

        // Iterate through the character arrays
        for (char c : chars1) {
            // Check if the character is a vowel
            if (isVowel(c)) {
                count1++;
            }
        }
        for (char c : chars2) {
            if (isVowel(c)) {
                count2++;
            }
        }
        // Return if the counts of vowels are equal
        return count1 == count2;
    }

    private boolean isVowel(char c) {
        // Convert the character to lowercase
        c = Character.toLowerCase(c);
        // Check if the character is a vowel
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
}